### ISSUE TITLE
New release does not fix startup problem

### ISSUE THREAD

**Issue by brhea** [juchong/ADIS16470-RoboRIO-Driver#15]
We loaded the new release yesterday and it did not solve the startup problem. Every time we power up, we have to then reboot the roboRio before the gyro will start. Values returned are 0.0 until we reboot. 

The problem seems a bit worse than before. Previously, when we deployed our Java code the gyro would start up every time after we deployed. Now, occasionally it does not start after deploying code which of course restarts the software.

**Comment by juchong** [comment]
@brhea The fix I pushed was only meant to address start-up issues with C++. I'll need more information in order to continue debugging your problem. Can you give the test robot project a try? I won't be able to help until I know how to reproduce the issue. Many teams are reporting that the driver works perfectly for them. https://github.com/juchong/ADIS16448-RoboRIO-Driver-Examples

**Comment by brhea** [comment]
I'll give it a try... I suppose I just swap the ADIS16470 object for the 16448 one. 
We didn't first notice this problem until we added the TalonSRX libraries (e.g. Phoenix) from CTRE.
It was shortly after that when we saw the problem so I'll add a talon after first testing without it. 
Thanks.

**Comment by ethanhmaness** [comment]
I found the example for the ADIS 16470 and cloned it. Set our team number properly and compiled and downloaded the program. It does start up and run properly part of the time. However, in about 2 out of 5 times, the gyro does not start. There is a slight difference than we've seen in our code. In the example, the gyro just freezes at a value.

Just to verify that we don't have network tables issue. I added a print statement to print out the gyro value:

  public void robotPeriodic() {

...
    System.out.println("YawAngle = " + m_imu.getAngle());

The gyro just freezes at a value and never changes. In our code it was always at 0.0. In your example... see below:

This is printed out every scan on the console... the number is exact.
YawAngle = -0.16472178809344765

I see exactly the same number in outline viewer and neither number changes....and moving the bot does not cause the number to change either.

The only change I made to your example was to add the print statement above to print to the console. 

We'll be testing another bot tomorrow... I think I've seen the same behavior on other roboRios. 

The problem exists without any Talon SRX's defined. Rebooting the roborio from the  drivers station fixes the problem and the gyro works fine.

**Comment by brhea** [comment]
I'm copying a post I made on Chief Delphi with a temporary solution to the root cause of this problem. I've not looked in detail at the buffer size in the acquire thread for your driver, but it appears that it is not large enough or the driver is not handling when that buffer size is exceeded. At least if the buffer needed is bigger than the one allocated, the code never recovers and the acquire thread never starts up properly.

The fix may not be the proper one but it may get us through this weekends competition. 


Our team is at Palmetto this weekend and the problem with the ADIS chip started happening every time that we power up our robot. The gyro might start functioning and then just stop (actually it failed every time we powered up our robot and rebooting the rio never fixed the problem). We found that errors were being thrown to the console indicating that a buffer had to be at least 4005 or 4080 , etc. after rebooting. The exact buffer requirement varied a bit on startup.

This was probably being caused because the startup time with the field system or when tethered is different that what we experienced when connecting wirelessly and the chip was sending more data than the buffer could handle during startup.

This is also why we probably started seeing the problem when we first added the CTRE phoenix libraries. They delayed startup a bit and enough data was being generated by the ADIS chip to overflow the buffer.

We had a look at the ADIS16470 driver and found the following in the acquire thread:

*/

private void acquire() {

// Set data packet length

final int dataset_len = 19; // 18 data points + timestamp

// Set up buffers and variables

int[] buffer = new int[4000];
A receive buffer of 4000 is created by the driver and the required size is bigger than 4000 under certain startup conditions.

We changed the buffer size to 10000 in the driver and the problem went away. The gyro now works on power up or any time we reboot our roboRio.

The problem seems to reside solely in the ADIS16470 driver. The problems we have witnessed were just exposed by loading the Phoenix library and also to a more severe extent when we disable wireless communication to the bot and tether with the competition configuration loaded in the radios.

**Comment by juchong** [comment]
I suspect that the issue is most prevalent when running additional user code alongside the driver. The 470 is configured to output data at 2000sps, which is completely unnecessary for FRC. Instead of increasing the buffer size, lowering the data output rate might provide better results. Can you try setting line 74 to the following: `WriteRegister(DEC_RATE, 0x0004);`? This should change the IMU output rate to 400Hz. Note that line 461 (`delta_angle = (ToInt(&buffer[i + 3]) * delta_angle_sf) / (500.0 / (buffer[i] - previous_timestamp));`) may need to be adjusted. I'll swing by your pit tomorrow morning to take a look. I'm one of your FTAs at Palmetto! What's your team number?

**Comment by brhea** [comment]
4020 Thanks

**Comment by brhea** [comment]
Thanks again for meeting with us in Palmetto.

The fix to increase the buffer was just a patch and not really a permanent or correct fix. When you look at fixing the error( we saw 1 instance where the 10K buffer size we used was not enough and we had to reboot to get the gyro to work). Consider the following potential change:

        data_count = m_spi.readAutoReceivedData(buffer, 0, 0); // Read number of bytes currently stored in the buffer
        data_remainder = data_count % dataset_len; // Check if frame is incomplete. Add 1 because of timestamp
        data_to_read = data_count - data_remainder; // Remove incomplete data from read count

        if (data_to_read > buffer.length) buffer = new int[data_to_read];  // reallocate if buffer will be exceeded
        
        m_spi.readAutoReceivedData(buffer, data_to_read, 0); // Read data from DMA buffer (only complete sets)


Checking the buffer to ensure that it is smaller than the data_to_read and reallocating the buffer size to accommodate the amount of data that will be read should keep the thread from crashing.... unless data_to_read was somehow huge and allocated all of the free memory in the roboRio.

Another potential fix would be to limit data_to_read to a maximum size and potentially loop through the reads a couple of time to get everything in the DMA buffer... 

data_to_read = data_count - data_remainder;

max_read = buffer.length()%dataset_len;   // get the integer size of max allowable read

data_to_read = Math.min(data_to_read, max_read);   // limit the read to be smaller than bufffer size


As you've suggested, slowing down the sampling rate might be a good idea also.... could some of the problem with drift be because the DMA buffer size is being exceed at times and some of the data is lost? 

Our autonomous relies on the gyro to be accurate in order to assure that we hit our targets and pick up balls from the correct location. We would not want to give up speed and sacrifice accuracy/repeatability.

**Comment by brhea** [comment]
I noted another place in the driver that may have the same bug as above. The switchToStandardSPI() method also reads data from the DMA buffer without limiting number of bytes read to the size of the receive buffer. This would also cause a thread crash... probably of the main robot loop instead of just the acquire thread.

Consider modifying to:

        int data_count = m_spi.readAutoReceivedData(trashBuffer, 0, 0);
        while (data_count > 0) {
          data_count = m_spi.readAutoReceivedData(trashBuffer, 0, 0);
          
          data_count = Math.min(trashBuffer.length, data_count);

          m_spi.readAutoReceivedData(trashBuffer, data_count, 0);
        }

**Comment by juchong** [comment]
@brhea, @ajn96 pushed an update to the c++ libraries that should help mitigate any overflow issues.

**Comment by brhea** [comment]
I made the same changes as the c++ changes to our copy of Java library. There seemed to be a problem with the scaling with the 400Hz rate. I changed the scale factor in the analog accumulator from 500 to 2500 just like in the c++ code and when we rotated our robot 90 degrees we only saw a 19 degree change in position.

I've set our local buffer size to 8200 to match the size of the DMA buffer and added the same checks you now have in C++ for buffer overrun and the driver seems to start and run properly. We went back to the higher DEC_RATE for now with a 500 scale factor.

**Comment by ajn96** [comment]
@brhea 
I took a look at the Java driver and discovered that there was a bug in the writeregister function preventing the lower half of each register from being written correctly. This made it so the decimation setting was never being applied (the IMU was still sampling at 2000sps but being scaled based on an expected 400sps (1/5 ratio on output)). 

This same issue was probably contributing to some of the calibration issues you were seeing, since the IMU null config register was not being set correctly. 

@juchong has fixed code and will be pushing an update shortly

**Comment by juchong** [comment]
@brhea The changes should be live. We just updated the maven repository with 2020.r4 which should resolve the issues you've been having

**Comment by juchong** [comment]
@brhea have you had a chance to test the updated libraries? Did they solve the issues you were having?

**Comment by brhea** [comment]
I plan to test changes this evening. I’ll send you a note with results.

**Comment by brhea** [comment]
I installed the update by checking online for updates( note that R4 installed instead of R3 which is listed as latest release). 

The driver seems to work properly. We tested by powering our robot down completely and restarting several times and the gyro did not fail to start. We also tested accuracy by aligning robot with a line and then turning 90 degrees to the line. This seemed to be pretty accurate. Final test of this will of course be using the Java driver with the FRC FMS and we can't do that until our next competition which is in Huntsville at the Rocket City Regional. 

I also noted that after calibration the drift seemed to be less than before but we still had our wrapper in place that rather quickly performs our own calibration rather soon after startup so we did not have much time to observe how much drift we continue to see. We'll also test the driver without our own calibration wrapper as time allows.
